src/System.Private.ServiceModel/src/System/ServiceModel/Security/SecurityUtils.cs

```diff
@@ -683,6 +779,54 @@ internal static string GetKeyDerivationAlgorithm(SecureConversationVersion versi
             }
         }
 
+        [SecuritySafeCritical]
+        internal static void FixNetworkCredential(ref NetworkCredential credential)
+        {
+            if (credential == null)
+            {
+                return;
+            }
+            string username = NetworkCredentialHelper.UnsafeGetUsername(credential);
+            string domain = NetworkCredentialHelper.UnsafeGetDomain(credential);
+            if (!string.IsNullOrEmpty(username) && string.IsNullOrEmpty(domain))
+            {
+                // do the splitting only if there is exactly 1 \ or exactly 1 @
+                string[] partsWithSlashDelimiter = username.Split('\\');
+                string[] partsWithAtDelimiter = username.Split('@');
+                if (partsWithSlashDelimiter.Length == 2 && partsWithAtDelimiter.Length == 1)
+                {
+                    if (!string.IsNullOrEmpty(partsWithSlashDelimiter[0]) && !string.IsNullOrEmpty(partsWithSlashDelimiter[1]))
+                    {
+                        credential = new NetworkCredential(partsWithSlashDelimiter[1], NetworkCredentialHelper.UnsafeGetPassword(credential), partsWithSlashDelimiter[0]);
+                    }
+                }
+                else if (partsWithSlashDelimiter.Length == 1 && partsWithAtDelimiter.Length == 2)
+                {
+                    if (!string.IsNullOrEmpty(partsWithAtDelimiter[0]) && !string.IsNullOrEmpty(partsWithAtDelimiter[1]))
+                    {
+                        credential = new NetworkCredential(partsWithAtDelimiter[0], NetworkCredentialHelper.UnsafeGetPassword(credential), partsWithAtDelimiter[1]);
+                    }
+                }
+            }
+        }
+
+        internal static void PrepareNetworkCredential()
+        {
+            if (s_dummyNetworkCredential == null)
+            {
+                PrepareNetworkCredentialWorker();
+            }
+        }
+
+        // Since this takes a lock, it probably won't be inlined, but the typical case will be.
+        static void PrepareNetworkCredentialWorker()
+        {
+            lock (s_dummyNetworkCredentialLock)
+            {
+                s_dummyNetworkCredential = new NetworkCredential("dummy", "dummy");
```

Comment:
On Desktop there is/was an issue when no NetworkCredential objects have ever been created yet and two threads tried to create a NetworkCredential object at the same time. There was some thread safety issue (I don't know exactly what the nature of the problem is). So the logic is such that we always create a dummy credential if it hasn't been created already, and do it inside a lock to ensure there's no problems. From then on, we can create a NetworkCredential object outside of a lock with no issues. This might or might not still be the case. I didn't want to spend the time to investigate if this is still an issue now as it would delay the feature.